pkg/bisect/bisect_test.go

```diff
@@ -36,46 +41,108 @@ func (env *testEnv) BuildKernel(compilerBin, userspaceDir, cmdlineFile, sysctlFi
 	return "", nil
 }
 
+func crashErrors(num int, title string) []error {
+	var errors []error
+	for i := 0; i < num; i++ {
+		errors = append(errors, &instance.CrashError{
+			Report: &report.Report{
+				Title: fmt.Sprintf("crashes at %v", title),
+			},
+		})
+	}
+	return errors
+}
+
+func nilErrors(num int) []error {
+	var errors []error
+	for i := 0; i < num; i++ {
+		errors = append(errors, nil)
+	}
+	return errors
+}
+
 func (env *testEnv) Test(numVMs int, reproSyz, reproOpts, reproC []byte) ([]error, error) {
 	hc, err := env.r.HeadCommit()
 	if err != nil {
 		env.t.Fatal(err)
 	}
-	// For cause bisection, if newer than or equal to 602, it crashes.
-	// -- 602 is the cause commit.
-	// TODO: for fix bisection(check env.fix), if older than 602, it crashes.
-	// -- 602 is the fix commit.
+
+	// Regardless of type of bisection, return that a crash occurs.
+	if env.crashAll {
+		return crashErrors(numVMs, hc.Title), nil
+	}
+
 	val, err := strconv.Atoi(hc.Title)
 	if err != nil {
 		env.t.Fatalf("invalid commit title: %v", val)
 	}
-	if val >= 602 {
-		var errors []error
-		for i := 0; i < numVMs; i++ {
-			errors = append(errors, &instance.CrashError{
-				Report: &report.Report{
-					Title: fmt.Sprintf("crashes at %v", hc.Title),
-				},
-			})
+
+	// Returns values that lead bisection to be inconclusive.
+	// Note that inconclusive results can also be caused by returning conflicting
+	// results, but we do not do that here.
+	// For cause bisection, let the causing commit be 602.
+	// -- [400, 500] => no crash
+	// -- [500, 700] => skip
+	// -- [700, HEAD] => crash
+	// -- 602 => crash
+	// For fix bisection, let the causing commit be 602.
+	// -- [400, 500] => crash
+	// -- [500, 700] => skip
+	// -- [700, HEAD] => no crash
+	// -- 602 => no crash
+	if env.inconclusive {
+		if !env.fix {
+			if val < 500 {
+				return nilErrors(numVMs), nil
+			} else if val > 700 {
+				return crashErrors(numVMs, hc.Title), nil
+			} else if val == 602 {
+				return crashErrors(numVMs, hc.Title), nil
+			} else if val > 500 && val < 700 {
+				return nilErrors(numVMs), fmt.Errorf("fake error")
+			}
+		} else {
+			if val < 500 {
+				return crashErrors(numVMs, hc.Title), nil
+			} else if val > 700 {
+				return nilErrors(numVMs), nil
+			} else if val == 602 {
```

Comment:
This looks overly complex to understand and inflexible as well (if one will want to add another case, it will be very hard to fit another case here without breaking all existing tests).

It seems that all of the cases can be encoded with 3 commits: culprit, brokenStart, brokenEnd. And we already have fix flag in the test. So this logic and be reduced to:
```
var res, err error
if commit >= brokenStart && commit <= brokenEnd {
	err = fmt.Errorf("broken build")
} else if commit < culprit == test.fix {
	res = instance.CrashError{Report: &report.Report{Title: fmt.Sprintf("crashes at %v", commit)}}	
}
```
Then the results will be driven from the test itself and it will be easy to add more tests without breaking existing ones.
No broken range can be encoded by setting brokenStart == brokenEnd == 0; still happens on oldest/newest release by setting culprit to 0/infinity, etc.

Reply:
Thanks! I've made the changes and updated the PR.